cluster: Sanction redpanda.leaders.preference on invalid license - create topic

IoannisRP opened the pull request:
Implements: [CORE-8042](https://redpandadata.atlassian.net/browse/CORE-8042) - Sanction 22

Implement sanction(s) on leadership pinning properties when there is no valid license:
- Requests to create a topic with a restricted `redpanda.leaders.preference` value will be denied.

## Backports Required

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v24.2.x
- [ ] v24.1.x
- [ ] v23.3.x

## Release Notes

* none

[CORE-8042]: https://redpandadata.atlassian.net/browse/CORE-8042?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

IoannisRP commented:
- rebased on top of #24065
- changed test helper to be type aware

BenPope commented on src/v/cluster/topics_frontend.cc:
```diff
@@ -81,6 +82,12 @@ std::vector<std::string_view> get_enterprise_features(
     if (cfg.is_schema_id_validation_enabled()) {
         features.emplace_back("schema ID validation");
     }
+    const config::leaders_preference no_preference{};
```
I wonder if we should reuse the default from the configuration?
```suggestion
    const auto no_preference{
      config::shard_local_cfg().default_leaders_preference.default_value()};

BenPope commented on src/v/kafka/server/tests/create_topics_test.cc:
```diff
@@ -447,27 +448,36 @@ FIXTURE_TEST(unlicensed_rejected, create_topic_fixture) {
-      ss::sstring{kafka::topic_property_record_value_schema_id_validation},
-      ss::sstring{
-        kafka::topic_property_record_value_schema_id_validation_compat},
+    using prop_t = std::map<ss::sstring, ss::sstring>;
+    const auto with = [](const std::string_view prop, const auto value) {
+        return std::make_pair(
+          prop, prop_t{{ss::sstring{prop}, ssx::sformat("{}", value)}});
     };
+    std::vector<std::pair<std::string_view, prop_t>> enterprise_props{
+      // si_props
+      with(kafka::topic_property_remote_read, true),
+      with(kafka::topic_property_remote_write, true),
+      with(kafka::topic_property_recovery, true),
+      with(kafka::topic_property_read_replica, true),
+      // schema id validation
+      with(kafka::topic_property_record_key_schema_id_validation, true),
+      with(kafka::topic_property_record_key_schema_id_validation_compat, true),
+      with(kafka::topic_property_record_value_schema_id_validation, true),
+      with(
+        kafka::topic_property_record_value_schema_id_validation_compat, true),
```
nitpick: It would be nice to split the refactor and the change into two commits.

BenPope reviewed (commented):
LGTM

IoannisRP commented:
- Rebased to dev
- split test refactoring into separate commit
- Replaced ad-hoc checking for restricted leadership_preference value with call to enterprise::check_restricted

michael-redpanda commented:
@ztlpn would appreciate your look here as well, thank you!

michael-redpanda commented on src/v/cluster/topics_frontend.cc:
```diff
@@ -81,6 +82,13 @@ std::vector<std::string_view> get_enterprise_features(
         features.emplace_back("schema ID validation");
     }
+    if (const auto leaders_pref = cfg.cfg.properties.leaders_preference;
+        leaders_pref
```
nit: I think our style is to explicitly check the result of `has_value()` on `std::optional`'s in conditions

BenPope commented on src/v/cluster/topics_frontend.cc:
```diff
@@ -81,6 +82,13 @@ std::vector<std::string_view> get_enterprise_features(
     if (cfg.is_schema_id_validation_enabled()) {
         features.emplace_back("schema ID validation");
     }
+    if (const auto leaders_pref = cfg.cfg.properties.leaders_preference;
```
nitpick: This doesn't need to be a copy.

IoannisRP replied:
you're right. There is a vector in there.

IoannisRP commented:
Addressed https://github.com/redpanda-data/redpanda/pull/24074#discussion_r1836720169 and https://github.com/redpanda-data/redpanda/pull/24074#discussion_r1836756053
